### PR TITLE
`azurerm_cognitive_account`: allows `bypass` property for `network_acls`

### LINKED ISSUE #28304: azurerm_cognitive_account - NetworkRuleSet support for bypass
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

I want to configure the following NetworkRuleSet https://learn.microsoft.com/en-us/rest/api/aiservices/accountmanagement/accounts/list?view=rest-aiservices-accountmanagement-2024-10-01&tabs=HTTP#networkruleset within the AzureRM resource `azurerm_cognitive_account` , but the property "bypass" is not yet available in the current version of the resource.

/https://github.com/hashicorp/terraform-provider-azurerm/pull/27851/ was recently merged that enable usage of the newer API version and reconfiguration of that resource is now possible.

### New or Affected Resource(s)/Data Source(s)

azurerm_cognitive_account

### Potential Terraform Configuration

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by feliperezende-barbosa**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR implements the `network_acls.bypass` property for `azurerm_coginitive_account`. [Azure Services for Azure OpenAI](https://learn.microsoft.com/en-us/azure/ai-services/cognitive-services-virtual-networks?tabs=portal#grant-access-to-trusted-azure-services-for-azure-openai)

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/cognitive -run=TestAccCognitiveAccount_networkAcls -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccCognitiveAccount_networkAclsVirtualNetworkRules
=== PAUSE TestAccCognitiveAccount_networkAclsVirtualNetworkRules
=== RUN   TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypass
=== PAUSE TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypass
=== RUN   TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypassKindNotSupported
=== PAUSE TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypassKindNotSupported
=== RUN   TestAccCognitiveAccount_networkAcls
=== PAUSE TestAccCognitiveAccount_networkAcls
=== CONT  TestAccCognitiveAccount_networkAclsVirtualNetworkRules
=== CONT  TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypassKindNotSupported
=== CONT  TestAccCognitiveAccount_networkAcls
=== CONT  TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypass
--- PASS: TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypassKindNotSupported (137.85s)
--- PASS: TestAccCognitiveAccount_networkAcls (216.38s)
--- PASS: TestAccCognitiveAccount_networkAclsVirtualNetworkRules (226.20s)
--- PASS: TestAccCognitiveAccount_networkAclsVirtualNetworkRulesWithBypass (258.18s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/cognitive	259.848s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_cognitive_account` - support for the `bypass` trusted Azure Services property [GH-[28221](https://github.com/hashicorp/terraform-provider-azurerm/pull/28221)]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #28304

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for this @feliperezende-barbos. This is looking good so far. Could we also add additional test steps to set this to `AzureServices` and back to `None` so we can confirm that updating this property works as expected? Thanks!

**Comment by feliperezende-barbosa** [comment]
> Thanks for this @feliperezende-barbos. This is looking good so far. Could we also add additional test steps to set this to `AzureServices` and back to `None` so we can confirm that updating this property works as expected? Thanks!

Sure, but just to add something. In the test I set the `bypass` to `None` first and then I update to `AzureServices`. Would you like to add one more step to back from `AzureServices` to `None` or just change the order of the acctest that already exists? Thanks.

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for updating the tests @feliperezende-barbosa. I had a look at the results and it appears we have a failure, but once that is addressed I can take another look. Thanks!

```
------- Stdout: -------
=== RUN   TestAccCognitiveAccount_networkAcls
=== PAUSE TestAccCognitiveAccount_networkAcls
=== CONT  TestAccCognitiveAccount_networkAcls
    testcase.go:173: Step 1/6 error: Error running apply: exit status 1
        Error: creating Account (Subscription: "*******"
        Resource Group Name: "acctestRG-cognitive-241219155302861835"
        Account Name: "acctestcogacc-241219155302861835"): unexpected status 400 (400 Bad Request) with error: NetworkAclsBypassNotSupported: The Kind 'Face' does not support Trusted Services.
          with azurerm_cognitive_account.test,
          on terraform_plugin_test.tf line 68, in resource "azurerm_cognitive_account" "test":
          68: resource "azurerm_cognitive_account" "test" {
--- FAIL: TestAccCognitiveAccount_networkAcls (177.11s)
FAIL
```

**Comment by feliperezende-barbosa** [comment]
> Thanks for updating the tests @feliperezende-barbosa. I had a look at the results and it appears we have a failure, but once that is addressed I can take another look. Thanks!
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccCognitiveAccount_networkAcls
> === PAUSE TestAccCognitiveAccount_networkAcls
> === CONT  TestAccCognitiveAccount_networkAcls
>     testcase.go:173: Step 1/6 error: Error running apply: exit status 1
>         Error: creating Account (Subscription: "*******"
>         Resource Group Name: "acctestRG-cognitive-241219155302861835"
>         Account Name: "acctestcogacc-241219155302861835"): unexpected status 400 (400 Bad Request) with error: NetworkAclsBypassNotSupported: The Kind 'Face' does not support Trusted Services.
>           with azurerm_cognitive_account.test,
>           on terraform_plugin_test.tf line 68, in resource "azurerm_cognitive_account" "test":
>           68: resource "azurerm_cognitive_account" "test" {
> --- FAIL: TestAccCognitiveAccount_networkAcls (177.11s)
> FAIL
> ```

I realized that you have tested the `TestAccCognitiveAccount_networkAcls` but the correct to test is `TestAccCognitiveAccount_networkAclsVirtualNetworkRules`. `TestAccCognitiveAccount_networkAcls` is for another sort of test. The problem was caused because the `bypass` property works for the Kind `OpenAI` and not for the Kind `Face` as you can see the error: `Account Name: "acctestcogacc-241219155302861835"): unexpected status 400 (400 Bad Request) with error: NetworkAclsBypassNotSupported: The Kind 'Face' does not support Trusted Services.` That's why I've changed the Kind from `Face` to `OpenAI` in `TestAccCognitiveAccount_networkAclsVirtualNetworkRules`. You can check it out in this [link](https://learn.microsoft.com/en-us/azure/ai-services/cognitive-services-virtual-networks?tabs=portal#grant-access-to-trusted-azure-services-for-azure-openai).

**Comment by catriona-m** [comment]
> > Thanks for updating the tests @feliperezende-barbosa. I had a look at the results and it appears we have a failure, but once that is addressed I can take another look. Thanks!
> > ```
> > ------- Stdout: -------
> > === RUN   TestAccCognitiveAccount_networkAcls
> > === PAUSE TestAccCognitiveAccount_networkAcls
> > === CONT  TestAccCognitiveAccount_networkAcls
> >     testcase.go:173: Step 1/6 error: Error running apply: exit status 1
> >         Error: creating Account (Subscription: "*******"
> >         Resource Group Name: "acctestRG-cognitive-241219155302861835"
> >         Account Name: "acctestcogacc-241219155302861835"): unexpected status 400 (400 Bad Request) with error: NetworkAclsBypassNotSupported: The Kind 'Face' does not support Trusted Services.
> >           with azurerm_cognitive_account.test,
> >           on terraform_plugin_test.tf line 68, in resource "azurerm_cognitive_account" "test":
> >           68: resource "azurerm_cognitive_account" "test" {
> > --- FAIL: TestAccCognitiveAccount_networkAcls (177.11s)
> > FAIL
> > ```
> 
> I realized you had tested the `TestAccCognitiveAccount_networkAcls` instead of `TestAccCognitiveAccount_networkAclsVirtualNetworkRules`. The problem was caused because the `bypass` property works for the Kind `OpenAI` and not for the Kind `Face` as you can see in the error: `Account Name: "acctestcogacc-241219155302861835"): unexpected status 400 (400 Bad Request) with error: NetworkAclsBypassNotSupported: The Kind 'Face' does not support Trusted Services.` That's why I've changed the Kind from `Face` to `OpenAI` in `TestAccCognitiveAccount_networkAclsVirtualNetworkRules`. You can check it out in this [link](https://learn.microsoft.com/en-us/azure/ai-services/cognitive-services-virtual-networks?tabs=portal#grant-access-to-trusted-azure-services-for-azure-openai). I fixed the `TestAccCognitiveAccount_networkAcls` because both tests are related. I've updated the PR in `Testing` section with the test results I've performed for both scenarios.

Thanks for your reply @feliperezende-barbosa. My concern is that adding this property seems to be affecting existing configurations which are not using the `bypass` property and which previously worked without error beforehand. Would you be able to take another look at fixing this up? Thanks!

**Comment by feliperezende-barbosa** [comment]
Thanks @catriona-m, it had the risk of affecting the existing configurations that are not using and do not support it. I've added more tests and one more validating an error that does not support the `bypass` property. Could you please take a look again? Thanks again.

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for updating this @feliperezende-barbosa - I have left one query inline but I can take another look once that is addressed. Thanks!

**Comment by feliperezende-barbosa** [comment]
> Thanks for updating this @feliperezende-barbosa - I have left one query inline but I can take another look once that is addressed. Thanks!

Hi @catriona-m, I cannot see any query inline. Could you please check this out? Thanks

File: internal/services/cognitive/cognitive_account_resource.go [hunk 1]
```diff
@@ -701,7 +717,17 @@ func expandCognitiveAccountNetworkAcls(d *pluginsdk.ResourceData) (*cognitiveser
 		IPRules:             &ipRules,
 		VirtualNetworkRules: &networkRules,
 	}
-	return &ruleSet, subnetIds
+
+	kind := d.Get("kind").(string)
+	if kind == "OpenAI" {
+		bypass := cognitiveservicesaccounts.ByPassSelection(v["bypass"].(string))
+		ruleSet.Bypass = &bypass
+	} else {
```
Comment (catriona-m) [inline internal/services/cognitive/cognitive_account_resource.go:725]: is it possible for a user to use a `kind` of `OpenAI` and not set `bypass`?

**Comment by catriona-m** [comment]
> > Thanks for updating this @feliperezende-barbosa - I have left one query inline but I can take another look once that is addressed. Thanks!
> 
> Hi @catriona-m, I cannot see any query inline. Could you please check this out? Thanks

Sorry about that @feliperezende-barbosa - not sure how it got lost, but Ive added the comment again now!

File: internal/services/cognitive/cognitive_account_resource.go (hunk 1 above)
Comment (feliperezende-barbosa) [reply internal/services/cognitive/cognitive_account_resource.go]: I have fixed this. Following when creating the resource on portal or using an API, the default value is `AzureServices`.

File: website/docs/r/cognitive_account.html.markdown [hunk 2]
```diff
@@ -98,6 +98,8 @@ The following arguments are supported:
 
 A `network_acls` block supports the following:
 
+* `bypass` - (Optional) Whether to allow trusted Azure Services to access the service. Possible values are `None` and `AzureServices`. Only the `Kind` of `OpenAI` is supported. Defaults to `AzureServices`.
+
```
Comment (catriona-m) [inline website/docs/r/cognitive_account.html.markdown:102]: There is no default value for `bypass` set in the schema currently, and if `bypass` can only be used when `kind` is `OpenAI`, then I don't think we should set a default for this property.

It would be good to add a note under the property about which kind is supported rather than having it in the description too.

```suggestion
* `bypass` - (Optional) Whether to allow trusted Azure Services to access the service. Possible values are `None` and `AzureServices`. 

-> **NOTE:** `bypass` can only be set when `kind` is set to `OpenAI `.

```

File: internal/services/cognitive/cognitive_account_resource.go [hunk 3]
```diff
@@ -701,7 +718,20 @@ func expandCognitiveAccountNetworkAcls(d *pluginsdk.ResourceData) (*cognitiveser
...
+	if kind == "OpenAI" {
+		if b, ok := d.GetOk("network_acls.0.bypass"); ok && b != "" {
+			bypass = cognitiveservicesaccounts.ByPassSelection(v["bypass"].(string))
+		}
+		ruleSet.Bypass = &bypass
+	} else {
+		if b, ok := d.GetOk("network_acls.0.bypass"); ok && b != "" {
+			return nil, nil, fmt.Errorf("the `network_acls.bypass` does not support Trusted Services for the kind %q", kind)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (feliperezende-barbosa) [reply internal/services/cognitive/cognitive_account_resource.go]: Much better. Thanks

**Review by catriona-m — APPROVED** [review]
Thanks for making the requested changes on this @feliperezende-barbosa - LGTM!
